api/v1beta1/cloudstackcluster_webhook_test.go

```diff
@@ -14,198 +14,85 @@ See the License for the specific language governing permissions and
-				cloudStackCluster = &CloudStackCluster{
-					TypeMeta: metav1.TypeMeta{
-						APIVersion: "infrastructure.cluster.x-k8s.io/v1beta1",
-						Kind:       "CloudStackCluster",
-					},
-					ObjectMeta: metav1.ObjectMeta{
-						Name:      "test-cluster2",
-						Namespace: clusterNamespace,
-					},
-					Spec: CloudStackClusterSpec{
-						IdentityRef: &CloudStackIdentityReference{
-							Kind: defaultIdentityRefKind,
-							Name: identitySecretName,
-						},
-						Zone:    zone,
-						Network: network,
-						// Need CP Endpoint not to be nil before test or webhook will allow modification.
-						ControlPlaneEndpoint: clusterv1.APIEndpoint{Host: "fakeIP", Port: int32(1234)},
-					},
-				}
-				cloudStackClusterUpdate = &CloudStackCluster{}
-			})
-
-			It("Should be rejected by the validating webhooks", func() {
-				Expect(k8sClient.Create(ctx, cloudStackCluster)).Should(Succeed())
-
-				forbiddenRegex := "admission webhook.*denied the request.*Forbidden\\: %s"
-
-				cloudStackCluster.DeepCopyInto(cloudStackClusterUpdate)
-				cloudStackClusterUpdate.Spec.Zone = "Zone2"
-				Expect(k8sClient.Update(ctx, cloudStackClusterUpdate).Error()).Should(MatchRegexp(forbiddenRegex, "zone"))
-
-				cloudStackCluster.DeepCopyInto(cloudStackClusterUpdate)
-				cloudStackClusterUpdate.Spec.Network = "Network2"
-				Expect(k8sClient.Update(ctx, cloudStackClusterUpdate).Error()).Should(MatchRegexp(forbiddenRegex, "network"))
-
-				cloudStackCluster.DeepCopyInto(cloudStackClusterUpdate)
-				cloudStackClusterUpdate.Spec.ControlPlaneEndpoint.Host = "1.1.1.1"
-				Expect(k8sClient.Update(ctx, cloudStackClusterUpdate).Error()).Should(MatchRegexp(forbiddenRegex, "controlplaneendpointhost"))
+		It("Should reject a CloudStackCluster with IdentityRef not of kind 'Secret'", func() {
+			dummies.CSCluster.Spec.IdentityRef.Kind = "ConfigMap"
+			Ω(k8sClient.Create(ctx, dummies.CSCluster)).
+				Should(MatchError(MatchRegexp(forbiddenRegex, "must be a Secret")))
+		})
+	})
 
-				cloudStackCluster.DeepCopyInto(cloudStackClusterUpdate)
-				cloudStackClusterUpdate.Spec.IdentityRef.Kind = "ConfigMap"
-				Expect(k8sClient.Update(ctx, cloudStackClusterUpdate).Error()).Should(MatchRegexp(forbiddenRegex, "identityRef\\.Kind"))
+	Context("When updating a CloudStackCluster", func() {
+		BeforeEach(func() {
+			Ω(k8sClient.Create(ctx, dummies.CSCluster)).Should(Succeed())
+		})
 
-				cloudStackCluster.DeepCopyInto(cloudStackClusterUpdate)
-				cloudStackClusterUpdate.Spec.IdentityRef.Name = configMapName
-				Expect(k8sClient.Update(ctx, cloudStackClusterUpdate).Error()).Should(MatchRegexp(forbiddenRegex, "identityRef\\.Name"))
-			})
+		It("Should reject updates to CloudStackCluster Zones", func() {
+			dummies.CSCluster.Spec.Zones = []infrav1.Zone{dummies.Zone1}
+			Ω(k8sClient.Update(ctx, dummies.CSCluster)).Should(MatchError(MatchRegexp(forbiddenRegex, "Zones and sub")))
+		})
+		It("Should reject updates to CloudStackCluster Zones", func() {
```

Comment:
Copy/paste from above?  It should reject updates to CloudStackCluster Zone's Networks.

Reply:
Ah, I should update the it statement.

In the first one I change the Zones entirely and in the second I change the network within a zone.

I'm working through a merge against main atm, but I'll update the description shortly.